### PR TITLE
[MCP] introduce ty::WhereClause to align chalk and rustc dyn repr

### LINKED ISSUE #30: align chalk and rustc dyn representation
chalk represents `dyn Trait` using binders, so that for `dyn Debug`, you have a predicate like `<T> T: Debug` (here, the bound variable `T` represents the self type).

rustc uses [`ExistentialPredicate`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/enum.ExistentialPredicate.html). 

The chalk way seems superior. Better to not have an extra form of predicate. 

The goal would be to refactor rustc to use a binder + predicate.

### LINKED ISSUE #433: Introduce ty::WhereClause to align chalk and rustc dyn representation
# Proposal

As @nikomatsakis wrote in https://github.com/rust-lang/wg-traits/issues/30 , chalk's binder + predicate representation seems better than rustc's [`ExistentialPredicate`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/enum.ExistentialPredicate.html).
https://github.com/rust-lang/rust/pull/85466 renamed `ExistentailPredicate` to `WhereClause`, and created a binder alias:
```rust
pub enum WhereClause {
    ...
}
pub type ExistentialPredicate<'tcx> = Binder<'tcx, WhereClause<'tcx>>;
```
# Mentors or Reviewers

@nikomatsakis 

# Process

The main points of the [Major Change Process][MCP] are as follows:

* [x] File an issue describing the proposal.
* [x] A compiler team member or contributor who is knowledgeable in the area can **second** by writing `@rustbot second`.
    * Finding a "second" suffices for internal changes. If however, you are proposing a new public-facing feature, such as a `-C flag`, then full team check-off is required.
    * Compiler team members can initiate a check-off via `@rfcbot fcp merge` on either the MCP or the PR.
* [ ] Once an MCP is seconded, the Final Comment Period begins. If no objections are raised after 10 days, the MCP is considered **approved**.

You can read [more about Major Change Proposals on forge][MCP].

[MCP]: https://forge.rust-lang.org/compiler/mcp.html

# Comments

**This issue is not meant to be used for technical discussion. There is a Zulip stream for that. Use this issue to leave procedural comments, such as volunteering to review, indicating that you second the proposal (or third, etc), or raising a concern that you would like to be addressed.**

### PR DESCRIPTION
**Pull request by csmoe**
Closes https://github.com/rust-lang/wg-traits/issues/30
Closes https://github.com/rust-lang/compiler-team/issues/433
r? @nikomatsakis

### CODE REVIEW

**Comment by csmoe** [comment]
@nikomatsakis ping for a review🔔

File: compiler/rustc_middle/src/ty/sty.rs [hunk 1]
```diff
@@ -716,11 +716,13 @@ pub enum ExistentialPredicate<'tcx> {
     AutoTrait(DefId),
 }
 
-impl<'tcx> ExistentialPredicate<'tcx> {
+pub type ExistentialPredicate<'tcx> = Binder<'tcx, WhereClause<'tcx>>;
```
Comment (jackh726) [inline compiler/rustc_middle/src/ty/sty.rs:719]: This isn't quite right. The difference between rustc and Chalk isn't solely the name. In Chalk, there's actually two layers of Binders. One that contains a single type that represents `Self`, and one that represents the vars on the predicates itself. See https://github.com/rust-lang/chalk/blob/802599ce2ca02f322c078f472e64c5cd3a208225/chalk-solve/src/rust_ir.rs#L233

**Review by jackh726 — CHANGES_REQUESTED** [review]

File: compiler/rustc_middle/src/ty/sty.rs (hunk 1 above)
Comment (jackh726) [reply compiler/rustc_middle/src/ty/sty.rs:719]: I guess the big thing missing here is the removal of `erase_self_ty`. And along with that, make the `Self` ty actually be a bound var.

**Comment by JohnCSimon** [comment]
ping from triage:
@csmoe 
Returning to you to address merge conflicts and after that switch this back to `S-waiting-on-review`.
thanks.

**Comment by nikomatsakis** [comment]
@csmoe sorry this PR disappeared from my radar

**Comment by nikomatsakis** [comment]
Are you still around?

**Comment by csmoe** [comment]
> Are you still around? 

yep, time was tough. I'm back :)

**Comment by nikomatsakis** [comment]
Augh! Let me give this a quick review @csmoe

**Review by nikomatsakis — COMMENTED** [review]
OK, so, I read the PR. From what I can tell, before this PR:

* We have `Binder<ExistentialPredicate>`, where `ExistentialPredicate` is a subset of predicates that don't have `Self` position specified.

After this PR:

* We introduce `WhereClause` to be the old `ExistentialPredicate`, and we make `ExistentialPredicate` an alias for `Binder<WC>`

This seems good, but indeed only a first step in the ultimate direction. Ultimately we want to extend `WhereClause` so it includes the `Self` type, and @jackh726 is right that this will mean introducing *another* level of Binder.

I think we also want to share WhereClause with Predicate.

**Comment by crlf0710** [comment]
@csmoe Ping from triage: Any updates here?

**Comment by Dylan-DPC** [comment]
Closing this as inactive

**Pull request closed without merging**